components/scream/src/physics/p3/p3_functions.hpp

```diff
@@ -0,0 +1,168 @@
+  using view_2d_table = typename KT::template view_2d_table<Scalar, Globals<ScalarT>::VTABLE_DIM0, Globals<ScalarT>::VTABLE_DIM1>;
+
+  template <typename S, int N>
+  using view_1d_ptr_array = typename KT::template view_1d_ptr_carray<S, N>;
+
+  using MemberType = typename KT::MemberType;
+
+  //
+  // --------- Functions ---------
+  //
+
+  // -- Table3
+
+  struct Table3 {
+    IntSmallPack dumii, dumjj;
+    Spack rdumii, rdumjj;
+  };
+
+  // Call from host to initialize the static table entries.
+  static void init_kokkos_tables(
+    view_2d_table& vn_table, view_2d_table& vm_table, view_1d_table& mu_r_table);
+
+  // Map (mu_r, lamr) to Table3 data.
+  KOKKOS_FUNCTION
+  static void lookup(const Smask& qr_gt_small, const Spack& mu_r, const Spack& lamr,
+                     Table3& t);
+
+  // Apply Table3 data to the table to return a value. This performs bilinear
+  // interpolation within the quad given by {t.dumii, t.dumjj} x {t.dumii+1,
+  // t.dumjj+1}.
+  KOKKOS_FUNCTION
+  static Spack apply_table(const Smask& qr_gt_small, const view_2d_table& table,
+                           const Table3& t);
+
+  // -- Sedimentation time step
+
+  // Calculate the first-order upwind step in the region [k_bot,
+  // k_top]. Velocity V is input, and flux is workspace and need not be
+  // initialized. On input, r contains mixing ratio data at the time step start;
+  // on output, it contains mixing ratio data at the time step end.
+  //
+  // A subtlety is that this procedure does not do exact upwind of a mixing
+  // ratio. That is because the background density rho is assumed to be static;
+  // rho does not get advected. Thus, there is an inconsistency between rho and
+  // r*rho at the level of |r|.
+
+  // Evolve nfield mixing ratios simultaneously. nfield is a compile-time
+  // parameter so the loops over nfield are compiled efficiently. So far the use
+  // cases have no need of a runtime version.
+  template <int nfield>
+  KOKKOS_FUNCTION
+  static void calc_first_order_upwind_step(
+    const ko::Unmanaged<view_1d<const Spack> >& rho,
+    const ko::Unmanaged<view_1d<const Spack> >& inv_rho, // 1/rho
+    const ko::Unmanaged<view_1d<const Spack> >& inv_dzq,
+    const MemberType& team,
+    const Int& nk, const Int& k_bot, const Int& k_top, const Int& kdir, const Scalar& dt_sub,
+    const view_1d_ptr_array<Spack, nfield>& flux, // workspace
+    const view_1d_ptr_array<Spack, nfield>& V,    // (behaviorally const)
+    const view_1d_ptr_array<Spack, nfield>& r);   // in/out
+
+  // Evolve 1 mixing ratio. This is a syntax-convenience version of the above.
+  KOKKOS_FUNCTION
+  static void calc_first_order_upwind_step(
+    const ko::Unmanaged<view_1d<const Spack> >& rho,
+    const ko::Unmanaged<view_1d<const Spack> >& inv_rho, // 1/rho
+    const ko::Unmanaged<view_1d<const Spack> >& inv_dzq,
+    const MemberType& team,
+    const Int& nk, const Int& k_bot, const Int& k_top, const Int& kdir, const Scalar& dt_sub,
```

Comment:
I see kdir a lot in the routines, and it appears to be used to switch between two different scenarios. Yet, I have no idea what it represent. Is there a more self-explanatory name for this variable? If not (e.g., it would be too long), can we add a quick explanation in the comment above (where some inputs/outputs are explained)?

Reply:
It represents the directions that columns are processed (bottom-up vs top-down). I'll add some comments.

Reply:
It basically enforces the direction for all the loops to go in the appropriate direction instead of from 1 to # of levels.  It's important for things like sedimentation, which should go from top to bottom.